src/sentry/utils/kvstore/bigtable.py

```diff
@@ -0,0 +1,287 @@
+import enum
+import logging
+import struct
+import zlib
+from datetime import timedelta
+from threading import Lock
+from typing import Any, Callable, Iterator, Mapping, NamedTuple, Optional, Sequence, Tuple, cast
+
+import zstandard
+from django.utils import timezone
+from google.api_core import exceptions, retry
+from google.cloud import bigtable
+from google.cloud.bigtable.row_data import PartialRowData
+from google.cloud.bigtable.row_set import RowSet
+from google.cloud.bigtable.table import Table
+from sentry.utils.kvstore.abstract import KVStorage
+
+
+logger = logging.getLogger(__name__)
+
+
+class CompressionStrategy(NamedTuple):
+    compress: Callable[[bytes], bytes]
+    decompress: Callable[[bytes], bytes]
+
+
+class BigtableError(Exception):
+    pass
+
+
+class BigtableKVStorage(KVStorage[str, bytes]):
+    max_size = 1024 * 1024 * 10
+    column_family = "x"
+
+    ttl_column = b"t"
+    ttl_struct = struct.Struct("<I")
+    flags_column = b"f"
+    flags_struct = struct.Struct("B")
+    data_column = b"0"
+
+    class Flags(enum.IntFlag):
+        # XXX: Compression flags are assumed to be mutually exclusive, the
+        # behavior is explicitly undefined if both bits are set on a record.
+        COMPRESSED_ZLIB = 1 << 0
+        COMPRESSED_ZSTD = 1 << 1
+
+    compression_strategies: Mapping[str, Tuple[Flags, CompressionStrategy]] = {
+        "zlib": (
+            Flags.COMPRESSED_ZLIB,
+            CompressionStrategy(compress=zlib.compress, decompress=zlib.decompress),
+        ),
+        "zstd": (
+            Flags.COMPRESSED_ZSTD,
+            CompressionStrategy(
+                compress=lambda x: cast(bytes, zstandard.ZstdCompressor().compress(x)),
+                decompress=lambda x: cast(bytes, zstandard.ZstdDecompressor().decompress(x)),
+            ),
+        ),
+    }
```

Comment:
This could also be `Mapping[Flags, CompressionStrategy]` but I thought having a `str` key for configuration's sake was a bit more ergonomic than passing a `Flag` from settings (the `Flag` also seems like it'd be likely to cause a circular import.)